File: 卡码/代码随想录/1/02-链表/Java/src/_06_链表相交/LeetCode_02_07.java
//面试题 02.07. 链表相交
package _06_链表相交;

public class LeetCode_02_07 {

}
class ListNode {
	int val;
	ListNode next;

	ListNode() {
	}

	ListNode(int val) {
		this.val = val;
	}

	ListNode(int val, ListNode next) {
		this.val = val;
		this.next = next;
	}
}
class Solution {
  public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
  	int lenA=0,lenB=0,gap=0;
  	ListNode tmpA=headA,tmpB=headB;
  	ListNode maxNode,minNode;
  	while(tmpA!=null) {
  		lenA++;
  		tmpA=tmpA.next;
  	}
  	while(tmpB!=null) {
  		lenB++;
  		tmpB=tmpB.next;
  	}
  	if(lenA>lenB) {
  		gap=lenA-lenB;
  		maxNode=headA;
  		minNode=headB;
  	}else {
  		gap=lenB-lenA;
  		maxNode=headB;
  		minNode=headA;
  	}
  	while(gap-->0) maxNode=maxNode.next;
  	while(maxNode!=null) {
  		if(maxNode==minNode) return maxNode;
  		maxNode=maxNode.next;
  		minNode=minNode.next;
  	}
  	return null;
  }
}

File: 卡码/代码随想录/1/02-链表/Go/06-链表相交/LeetCode_02_07.go
//面试题 02.07. 链表相交
package main

type ListNode struct{
	Val int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	lenA,lenB:=0,0	//A,B长度
	tmpA,tmpB:=headA,headB
	for tmpA!=nil {
		lenA++
		tmpA=tmpA.Next
	}
	for tmpB!=nil {
		lenB++
		tmpB=tmpB.Next
	}
	var gap int	//最大长度与最小长度之差
	var maxNode,minNode *ListNode	//最大、最小长度对应的链表
	if lenA>lenB {
		gap=lenA-lenB
		maxNode=headA
		minNode=headB
	}else{
		gap=lenB-lenA
		maxNode=headB
		minNode=headA
	}
	for gap>0 {
		maxNode=maxNode.Next
		gap--
	}
	for maxNode!=nil {
		if maxNode==minNode {
			return maxNode
		}
		maxNode=maxNode.Next
		minNode=minNode.Next
	}
	return nil
}
